File: src/old/quest_editor/scripting/disassembly.ts
import { reinterpret_i32_as_f32 } from "../../core/primitive_conversion";
import { Arg, Segment, SegmentType } from "./instructions";
import { Kind, Param, StackInteraction } from "./opcodes";

/**
 * @param manual_stack If true, will output stack management instructions (argpush variants). Otherwise the arguments of stack management instructions will be output as arguments to the instruction that pops them from the stack.
 */
export function disassemble(object_code: Segment[], manual_stack: boolean = false): string[] {
    const lines: string[] = [];
    const stack: Arg[] = [];
    let section_type: SegmentType | undefined;

    for (const segment of object_code) {
        // Section marker.
        let section_marker!: string;

        switch (segment.type) {
            case SegmentType.Instructions:
                section_marker = ".code";
                break;
            case SegmentType.Data:
                section_marker = ".data";
                break;
            case SegmentType.String:
                section_marker = ".string";
                break;
        }

        if (section_type !== segment.type) {
            section_type = segment.type;

            if (lines.length) {
                lines.push("");
            }

            lines.push(section_marker, "");
        }

        // Labels.
        for (const label of segment.labels) {
            lines.push(`${label}:`);
        }

        // Code or data lines.
        if (segment.type === SegmentType.Data) {
            const bytes = new Uint8Array(segment.data);
            let line = "    ";

            for (let i = 0; i < bytes.length; i++) {
                line += "0x" + bytes[i].toString(16).padStart(2, "0");

                if (i % 16 === 15) {
                    lines.push(line);
                    line = "    ";
                } else if (i < bytes.length - 1) {
                    line += " ";
                }
            }

            if (line.length > 4) {
                lines.push(line);
            }
        } else if (segment.type === SegmentType.String) {
            lines.push("    " + JSON.stringify(segment.value));
        } else {
            for (const instruction of segment.instructions) {
                if (!manual_stack && instruction.opcode.stack === StackInteraction.Push) {
                    stack.push(...instruction.args);
                } else {
                    let args: string[] = [];

                    if (instruction.opcode.stack === StackInteraction.Pop) {
                        if (!manual_stack) {
                            args = args_to_strings(
                                instruction.opcode.params,
                                stack.splice(
                                    Math.max(0, stack.length - instruction.opcode.params.length),
                                    instruction.opcode.params.length,
                                ),
                                true,
                            );
                        }
                    } else {
                        args = args_to_strings(instruction.opcode.params, instruction.args, false);
                    }

                    lines.push(
                        "    " +
                            instruction.opcode.mnemonic +
                            (args.length ? " " + args.join(", ") : ""),
                    );
                }
            }
        }
    }

    // Ensure newline at the end.
    if (lines.length) {
        lines.push("");
    }

    return lines;
}

function args_to_strings(params: Param[], args: Arg[], stack: boolean): string[] {
    const arg_strings: string[] = [];

    for (let i = 0; i < params.length; i++) {
        const type = params[i].type;
        const arg = args[i];

        if (arg == null) {
            arg_strings.push("");
            continue;
        }

        switch (type.kind) {
            case Kind.Float:
                // Floats are pushed onto the stack as integers with arg_pushl.
                if (stack) {
                    arg_strings.push(reinterpret_i32_as_f32(arg.value).toString());
                } else {
                    arg_strings.push(arg.value.toString());
                }
                break;
            case Kind.ILabelVar:
                for (; i < args.length; i++) {
                    arg_strings.push(args[i].value.toString());
                }
                break;
            case Kind.RegRefVar:
                for (; i < args.length; i++) {
                    arg_strings.push("r" + args[i].value);
                }
                break;
            case Kind.RegRef:
            case Kind.RegTupRef:
                arg_strings.push("r" + arg.value);
                break;
            case Kind.String:
                arg_strings.push(JSON.stringify(arg.value));
                break;
            default:
                arg_strings.push(arg.value.toString());
                break;
        }
    }

    return arg_strings;
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/asm/Disassembly.kt
package world.phantasmal.lib.asm

import mu.KotlinLogging
import kotlin.math.min

private val logger = KotlinLogging.logger {}

private const val INDENT_WIDTH = 4
private val INDENT = " ".repeat(INDENT_WIDTH)

/**
 * @param inlineStackArgs If true, will output stack arguments inline instead of outputting stack
 * management instructions (argpush variants).
 */
fun disassemble(bytecodeIr: BytecodeIr, inlineStackArgs: Boolean = true): List<String> {
    logger.trace {
        "Disassembling ${bytecodeIr.segments.size} segments with ${
            if (inlineStackArgs) "inline stack arguments" else "stack push instructions"
        }."
    }

    val lines = mutableListOf<String>()
    val stack = mutableListOf<ArgWithType>()
    var sectionType: SegmentType? = null

    for (segment in bytecodeIr.segments) {
        // Section marker (.code, .data or .string).
        if (sectionType != segment.type) {
            sectionType = segment.type

            if (lines.isNotEmpty()) {
                lines.add("")
            }

            val sectionMarker = when (segment) {
                is InstructionSegment -> ".code"
                is DataSegment -> ".data"
                is StringSegment -> ".string"
            }

            lines.add(sectionMarker)
            lines.add("")
        }

        // Labels.
        for (label in segment.labels) {
            lines.add("$label:")
        }

        // Code or data lines.
        when (segment) {
            is InstructionSegment -> {
                var inVaList = false

                segment.instructions.forEachIndexed { i, instruction ->
                    val opcode = instruction.opcode

                    if (opcode.code == OP_VA_START.code) {
                        inVaList = true
                    } else if (opcode.code == OP_VA_END.code) {
                        inVaList = false
                    }

                    if (inlineStackArgs &&
                        !inVaList &&
                        opcode.stack == StackInteraction.Push &&
                        canInlinePushedArg(segment, i)
                    ) {
                        stack.addAll(addTypeToArgs(opcode.params, instruction.args))
                    } else {
                        val sb = StringBuilder(INDENT)
                        sb.append(opcode.mnemonic)

                        if (opcode.stack == StackInteraction.Pop) {
                            if (inlineStackArgs) {
                                sb.appendArgs(
                                    opcode.params,
                                    stack.takeLast(opcode.params.size),
                                    stack = true,
                                )
                            }
                        } else {
                            sb.appendArgs(
                                opcode.params,
                                addTypeToArgs(opcode.params, instruction.args),
                                stack = false
                            )
                        }

                        if (opcode.stack != StackInteraction.Push) {
                            stack.clear()
                        }

                        lines.add(sb.toString())
                    }
                }
            }

            is DataSegment -> {
                val sb = StringBuilder(INDENT)

                for (i in 0 until segment.data.size) {
                    sb.append("0x")
                    sb.append(segment.data.getUByte(i).toString(16).padStart(2, '0'))

                    when {
                        // Last line.
                        i == segment.data.size - 1 -> {
                            lines.add(sb.toString())
                        }
                        // Start a new line after every 16 bytes.
                        i % 16 == 15 -> {
                            lines.add(sb.toString())
                            sb.setLength(0)
                            sb.append(INDENT)
                        }
                        // Add a space between each byte.
                        else -> {
                            sb.append(" ")
                        }
                    }
                }
            }

            is StringSegment -> {
                lines.add(StringBuilder(INDENT).appendStringSegment(segment.value).toString())
            }
        }
    }

    // Ensure newline at the end.
    lines.add("")

    logger.trace { "Disassembly finished, line count: ${lines.size}." }

    return lines
}

private data class ArgWithType(val arg: Arg, val type: AnyType)

private fun canInlinePushedArg(segment: InstructionSegment, index: Int): Boolean {
    var pushedArgCount = 0

    for (i in index until segment.instructions.size) {
        val opcode = segment.instructions[i].opcode

        when (opcode.stack) {
            StackInteraction.Push -> pushedArgCount++

            StackInteraction.Pop -> {
                var paramCount = 0
                var varArgs = false

                for (param in opcode.params) {
                    when (param.type) {
                        is ILabelVarType -> varArgs = true
                        is RegVarType -> varArgs = true
                        else -> paramCount++
                    }
                }

                return pushedArgCount <= paramCount || (pushedArgCount > paramCount && varArgs)
            }

            null -> return false
        }
    }

    return false
}

private fun addTypeToArgs(params: List<Param>, args: List<Arg>): List<ArgWithType> {
    val argsWithType = mutableListOf<ArgWithType>()

    for (i in 0 until min(params.size, args.size)) {
        argsWithType.add(ArgWithType(args[i], params[i].type))
    }

    // Deal with varargs.
    val lastParam = params.lastOrNull()

    if (lastParam?.varargs == true) {
        for (i in argsWithType.size until args.size) {
            argsWithType.add(ArgWithType(args[i], lastParam.type))
        }
    }

    return argsWithType
}

private fun StringBuilder.appendArgs(params: List<Param>, args: List<ArgWithType>, stack: Boolean) {
    var i = 0

    while (i < params.size) {
        val paramType = params[i].type

        if (i == 0) {
            append(" ")
        } else {
            append(", ")
        }

        if (i < args.size) {
            val (arg, argType) = args[i]

            if (argType is RegType) {
                append("r")
                append(arg.value)
            } else {
                when (paramType) {
                    FloatType -> {
                        // Floats are pushed onto the stack as integers with arg_pushl.
                        if (stack) {
                            append(Float.fromBits((arg as IntArg).value))
                        } else {
                            append(arg.value)
                        }
                    }

                    ILabelVarType -> {
                        while (i < args.size) {
                            append(args[i].arg.value)
                            if (i < args.lastIndex) append(", ")
                            i++
                        }
                    }

                    RegVarType -> {
                        while (i < args.size) {
                            append("r")
                            append(args[i].arg.value)
                            if (i < args.lastIndex) append(", ")
                            i++
                        }
                    }

                    is RegType -> {
                        append("r")
                        append(arg.value)
                    }

                    StringType -> {
                        appendStringArg((arg as StringArg).value)
                    }

                    else -> {
                        append(arg.value)
                    }
                }
            }
        }

        i++
    }
}

private fun StringBuilder.appendStringArg(value: String): StringBuilder {
    append("\"")

    for (char in value) {
        when (char) {
            '\r' -> append("\\r")
            '\n' -> append("\\n")
            '\t' -> append("\\t")
            '"' -> append("\\\"")
            else -> append(char)
        }
    }

    append("\"")
    return this
}

private fun StringBuilder.appendStringSegment(value: String): StringBuilder {
    append("\"")

    var i = 0

    while (i < value.length) {
        when (val char = value[i]) {
            // Replace <cr> with \n.
            '<' -> {
                if (i + 3 < value.length &&
                    value[i + 1] == 'c' &&
                    value[i + 2] == 'r' &&
                    value[i + 3] == '>'
                ) {
                    append("\\n")
                    i += 3
                } else {
                    append(char)
                }
            }
            '\r' -> append("\\r")
            '\n' -> append("\\n")
            '\t' -> append("\\t")
            '"' -> append("\\\"")
            else -> append(char)
        }

        i++
    }

    append("\"")
    return this
}
